src/dotnet/Telemetry/TelemetryCommonProperties.cs

```diff
@@ -73,5 +89,135 @@ private string GetMachineId()
+        private static extern bool GetProductInfo(uint dwOSMajorVersion, uint dwOSMinorVersion, uint dwSpMajorVersion, uint dwSpMinorVersion, out uint pdwReturnedProductType);
+
+        /// <summary>
+        /// For Windows, returns the product type, loosely the SKU, as encoded by GetProductInfo().
+        /// For example, Enterprise is "4" (0x4) and Professional is "48" (0x30)
+        /// See https://msdn.microsoft.com/en-us/library/windows/desktop/ms724358(v=vs.85).aspx for the full list.
+        /// We're not attempting to decode the value on the client side as new Windows releases may add new values.
+        /// For Unix, currently returns "4294967295" (uint.MaxValue).
+        /// </summary>
+        /// <returns></returns>
+        private static string GetProductType()
+        {
+            if (RuntimeEnvironment.OperatingSystemPlatform != Platform.Windows)
+                return "0";
+
+            if (!GetProductInfo((uint)Environment.OSVersion.Version.Major, (uint)Environment.OSVersion.Version.Minor, 0, 0, out uint productType))
+            {
+                return uint.MaxValue.ToString("D"); // Error
+            }
+
+            return productType.ToString("D");
+        }
+
+        [DllImport("libc", ExactSpelling = true, CharSet = CharSet.Ansi, CallingConvention = CallingConvention.Cdecl)]
+        private static extern IntPtr gnu_get_libc_release();
+
+        [DllImport("libc", ExactSpelling = true, CharSet = CharSet.Ansi, CallingConvention = CallingConvention.Cdecl)]
+        private static extern IntPtr gnu_get_libc_version();
+
+        /// <summary>
+        /// If gnulibc is available, returns the release, such as "stable".
+        /// If the libc is musl, currently returns empty string.
+        /// Otherwise returns empty string.
+        /// </summary>
+        private static string GetLibcRelease()
+        {
+            if (RuntimeEnvironment.OperatingSystemPlatform == Platform.Windows)
+                return "";
+
+            try
+            {
+                return Marshal.PtrToStringUTF8(gnu_get_libc_release());
+            }
+            catch (Exception e) when (e is DllNotFoundException | e is EntryPointNotFoundException)
+            {
+                return "";
+            }
+        }
+
+        /// <summary>
+        /// If gnulibc is available, returns the version, such as "2.22".
+        /// If the libc is musl, currently returns empty string. (In future could run "ldd -version".)
+        /// Otherwise returns empty string.
+        /// </summary>
+        private static string GetLibcVersion()
+        {
+            if (RuntimeEnvironment.OperatingSystemPlatform == Platform.Windows)
+                return "";
+
+            try
+            {
+                return Marshal.PtrToStringUTF8(gnu_get_libc_version());
+            }
+            catch (Exception e) when (e is DllNotFoundException | e is EntryPointNotFoundException)
```

Comment:
Let me know if I should replace the catches with blanket catches. We should not fall over for telemetry failure.

Reply:
done